---
title: "Figure 4"
author: "Jasper Panten"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Desktop/Projects/ASE_Spermatogenesis_Paper_rerun_Revisions//")
```

## This script reproduces the plots shown in figure 4
## Dynamic expression divergence between mouse sub-species is predominantly caused by differentiation-dependent cis-effects

# Analysis feeding into this figure:
./Figure4_gp_fits_de.R

```{r code}

setwd("~/Desktop/Projects/ASE_Spermatogenesis_Paper_rerun_Revisions//")

source("./Scripts/General/auxiliary.R")
source("./Scripts/General/reuse_functions.R")

library(ggplot2)
library(pheatmap)
library(ComplexHeatmap)
library(tidyverse)

results_per_gene_expressed_intervals <- readRDS("./Data/processed/trans_effects_expressed_intervals_new.rds")

genes_test <- readRDS("./Data/processed/genes_test.rds")
genes_save <- rownames(genes_test)
genes_test <- data.frame(apply(genes_test, 2, as.numeric))
rownames(genes_test) <- genes_save
genes_test <- genes_test[!is.na(rownames(genes_test)), ]
genes_test <- genes_test[!is.na(genes_test$chromosome), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans), ]
genes_test <- genes_test[!is.na(genes_test$BF_cis_dynamic), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans_static), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans_dynamic), ]

# add static cis BF as NLL_null - NLL_alt from dali LLR test
genes_test$BF_cis_static <- genes_test$NLL_null - genes_test$NLL_alt
genes_test[genes_test$NLL_null < 0 | genes_test$NLL_alt < 0, ]$BF_cis_static <- 0
genes_test[genes_test$BF_cis_static < 0, ]$BF_cis_static <- 0

data_use <- readRDS("./Data/processed/sce_merged_new.rds")
total_exp <- readRDS("./Data/processed/total_exp.rds")

pseudotime_here <- colData(data_use)$Pseudotime

genes_cis_static <- rownames(genes_test[genes_test$BF_cis_static > 10 & genes_test$BF_cis_dynamic < 10, ])
genes_cis_dynamic <- rownames(genes_test[genes_test$BF_cis_dynamic > 10, ])
genes_trans_static <- rownames(genes_test[genes_test$BF_trans_static > 10 & genes_test$BF_trans_dynamic < 10, ])
genes_trans_dynamic <- rownames(genes_test[genes_test$BF_trans_dynamic > 10, ])

counts_reference <- function(sce){
  assays(sce)[["counts_reference"]]
}
counts_alternative <- function(sce){
  assays(sce)[["counts_alternative"]]
}

# correlation coefficients between cis (f1) and trans (f0 - f1)

genes_use <- rownames(genes_test)
chromosome_genes_use <- rowData(data_use[rownames(genes_test), ])$chromosome_name
chromosome_genes_use[is.na(chromosome_genes_use)] <- "NA"
table(chromosome_genes_use)
genes_use <- genes_use[!chromosome_genes_use %in% c("MT", "X")]

genes_test_here <- genes_test

df_dynamics <- data.frame(
  genes = rownames(genes_test_here), 
  cis_static = genes_test_here$NLL_null - genes_test_here$NLL_alt, 
  cis_dynamics = genes_test_here$BF_cis_dynamic, 
  trans_dynamics = genes_test_here$BF_trans_dynamic, 
  trans_static = genes_test_here$BF_trans_static, 
  trans_total = genes_test_here$BF_trans
)
df_dynamics[df_dynamics$cis_static < 0, ]$cis_static <- 0

test_static <- fisher.test(table(df_dynamics$cis_dynamics > 10, df_dynamics$trans_static > 10))
test_dynamic <- fisher.test(table(df_dynamics$cis_dynamics > 10, df_dynamics$trans_dynamics > 10))

df_dynamics[df_dynamics$cis_static > 100, ]$cis_static <- 100
df_dynamics[df_dynamics$trans_static > 25, ]$trans_static <- 25
df_dynamics[df_dynamics$cis_dynamics > 50, ]$cis_dynamics <- 50
df_dynamics[df_dynamics$trans_dynamics > 25, ]$trans_dynamics <- 25

ggplot(df_dynamics, aes(cis_static, trans_static)) + 
  stat_density2d(geom = "raster", mapping = aes(fill = after_stat(density)), contour = FALSE, h = 20) + 
  scale_fill_gradient(low = "white", high = "blue") + 
  geom_jitter(col = "black", width = 1, height = 1, size = 2, alpha = 1) + 
  geom_vline(xintercept = 10, linetype = "dashed") + geom_hline(yintercept = 10, linetype = "dashed") + 
  scale_color_gradient2(mid = "darkgrey") + theme_classic()  + 
  #ylim(-2, 30) + 
  scale_x_continuous(limits = c(-1, 105), expand = c(0.02, 0.02), breaks = c(0, 10, 50, 100)) +
  scale_y_continuous(limits = c(-1, 27), expand = c(0.02, 0.02)) +
  theme(text = element_text(size = 30)) + 
  xlab("Cis (logBF)") + 
  ylab("Trans (logBF)") + 
  ggtitle(paste0("Persistent effects (", paste0("OR = ", round(test_static$estimate, digits = 2)), ")")) + 
  theme(legend.position = "None")
ggsave("./Plots/Figure4/cis_trans_static.pdf", width = 20, height = 20, units = "cm")

ggplot(df_dynamics, aes(cis_dynamics, trans_dynamics)) + 
  stat_density2d(geom = "raster", mapping = aes(fill = after_stat(density)), contour = FALSE, h = 20) + 
  scale_fill_gradient(low = "white", high = "blue") + 
  geom_jitter(col = "black", width = 1, height = 0, size = 3, alpha = 1) + 
  geom_vline(xintercept = 10, linetype = "dashed") + geom_hline(yintercept = 10, linetype = "dashed") + 
  scale_color_gradient2(mid = "darkgrey") + theme_classic()  + 
  #ylim(-2, 30) + 
  scale_x_continuous(limits = c(-1, 55), expand = c(0.02, 0.02), breaks = c(0, 10, 25, 50)) +
  scale_y_continuous(limits = c(-1, 27), expand = c(0.02, 0.02)) +
  theme(text = element_text(size = 30)) + 
  xlab("Cis (logBF)") + 
  ylab("Trans (logBF)") + 
  ggtitle(paste0("Dynamic effects (", paste0("OR = ", round(test_dynamic$estimate, digits = 2)), ")")) + 
  theme(legend.position = "None")
ggsave("./Plots/Figure4/cis_trans_dynamics.pdf", width = 20, height = 20, units = "cm")

## 

make_meta_profiles <- function(dataset, k = 5){
  dis_mat <- dist(dataset)
  expression_clusters <- cutree(hclust(dis_mat, method = "complete"), k = k)
  cluster_counts <- table(expression_clusters)
  print(cluster_counts)
  res <- lapply(unique(expression_clusters), function(i){
    dd <- dataset[expression_clusters == i, ]
    if (  cluster_counts[[i]] == 1 ){
      return(dd)
    }
    return(colMeans(dd, na.rm = T))
  })
  return(do.call("rbind", res))
}

# load all the data

library(ggplot2)
library(pheatmap)
library(ComplexHeatmap)
library(tidyverse)

results_per_gene_expressed_intervals <- readRDS("./Data/processed/trans_effects_expressed_intervals_new.rds")

genes_test <- readRDS("./Data/processed/genes_test.rds")
genes_save <- rownames(genes_test)
genes_test <- data.frame(apply(genes_test, 2, as.numeric))
rownames(genes_test) <- genes_save
genes_test <- genes_test[!is.na(rownames(genes_test)), ]
genes_test <- genes_test[!is.na(genes_test$chromosome), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans), ]
genes_test <- genes_test[!is.na(genes_test$BF_cis_dynamic), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans_static), ]
genes_test <- genes_test[!is.na(genes_test$BF_trans_dynamic), ]

# add static cis BF as NLL_null - NLL_alt from dali LLR test
# there are some issues here with negative NLLs resulting from model misspecification, should be fixed in the new model
genes_test$BF_cis_static <- genes_test$NLL_null - genes_test$NLL_alt
genes_test[genes_test$NLL_null < 0 | genes_test$NLL_alt < 0, ]$BF_cis_static <- 0
genes_test[genes_test$BF_cis_static < 0, ]$BF_cis_static <- 0

genes_cis_static <- rownames(genes_test[genes_test$BF_cis_static > 10 & genes_test$BF_cis_dynamic < 10, ])
genes_cis_dynamic <- rownames(genes_test[genes_test$BF_cis_dynamic > 10, ])
genes_trans_static <- rownames(genes_test[genes_test$BF_trans_static > 10 & genes_test$BF_trans_dynamic < 10, ])
genes_trans_dynamic <- rownames(genes_test[genes_test$BF_trans_dynamic > 10, ])

data_use <- readRDS("./Data/processed/sce_merged_new.rds")
total_exp <- readRDS("./Data/processed/total_exp.rds")

counts_reference <- function(sce){
  assays(sce)[["counts_reference"]]
}
counts_alternative <- function(sce){
  assays(sce)[["counts_alternative"]]
}

pseudotime_here <- colData(data_use)$Pseudotime
#pseudotime_here <- - (pseudotime_here - max(pseudotime_here))

make_pseudotime_smooth <- function(data, pt, n_partitions = 100){
  pseudotime_here <- pt
  #pseudotime_here <- pseudotime_here - min(pseudotime_here)
  pseudotime_here <- pseudotime_here / max(pseudotime_here) # between 0 and 1
  intervals <- c(0, 1:n_partitions / n_partitions)
  
  expression_smoothed <- do.call("cbind", 
                                 lapply(1:n_partitions, function(i){
                                   p_lower <- intervals[i]
                                   p_upper <-  intervals[i + 1]
                                   ix <- which(p_lower <= pseudotime_here & pseudotime_here <= p_upper)
                                   rowMeans(data[,ix])
                                 }))
  expression_smoothed
}

data_f0_b6_rep1 <- make_pseudotime_smooth(counts_reference(data_use[,data_use$Library == "Sample1"]),
                                          pseudotime_here[data_use$Library == "Sample1"], n_partitions = 100)
data_f0_b6_rep2 <- make_pseudotime_smooth(counts_reference(data_use[,data_use$Library == "Sample2"]),
                                          pseudotime_here[data_use$Library == "Sample2"], n_partitions = 100)
data_f0_cast_rep1 <- make_pseudotime_smooth(counts_alternative(data_use[,data_use$Library == "Sample3"]),
                                            pseudotime_here[data_use$Library == "Sample3"], n_partitions = 100)
data_f0_cast_rep2 <- make_pseudotime_smooth(counts_alternative(data_use[,data_use$Library == "Sample4"]),
                                            pseudotime_here[data_use$Library == "Sample4"], n_partitions = 100)
data_f1_b6_rep1 <- make_pseudotime_smooth(counts_reference(data_use[,data_use$Library == "Sample5"]),
                                          pseudotime_here[data_use$Library == "Sample5"], n_partitions = 100)
data_f1_b6_rep2 <- make_pseudotime_smooth(counts_reference(data_use[,data_use$Library == "Sample6"]),
                                          pseudotime_here[data_use$Library == "Sample6"], n_partitions = 100)
data_f1_cast_rep1 <- make_pseudotime_smooth(counts_alternative(data_use[,data_use$Library == "Sample5"]),
                                            pseudotime_here[data_use$Library == "Sample5"], n_partitions = 100)
data_f1_cast_rep2 <- make_pseudotime_smooth(counts_alternative(data_use[,data_use$Library == "Sample6"]),
                                            pseudotime_here[data_use$Library == "Sample6"], n_partitions = 100)

saveRDS("./Data/processed/binned_data.rds")
list_data <- readRDS("./Data/processed/binned_data.rds")

data_f0_b6 <- list_data[[1]]
data_f0_cast <- list_data[[2]]
data_f1_b6 <- list_data[[3]]
data_f1_cast <- list_data[[4]]

sfs = c(sum(data_f0_b6), sum(data_f0_cast)) / sum(data_f0_b6)
data_f0_b6_norm <- data_f0_b6 / sfs[[1]]
data_f0_cast_norm <- data_f0_cast / sfs[[2]]

genes_use <- rownames(data_f0_b6_norm)
rs_f0 <- (data_f0_b6_norm[genes_use, ] + 1) / (data_f0_b6_norm[genes_use, ] + data_f0_cast_norm[genes_use, ] + 2)
rs_f1 <- (data_f1_b6[genes_use, ] + 1) / (data_f1_b6[genes_use, ] + data_f1_cast[genes_use, ] + 2)

rs_f0_no_offset <- (data_f0_b6_norm[genes_use, ]) / (data_f0_b6_norm[genes_use, ] + data_f0_cast_norm[genes_use, ])
rs_f1_no_offset <- (data_f1_b6[genes_use, ]) / (data_f1_b6[genes_use, ] + data_f1_cast[genes_use, ])

total_expression <- data_f0_b6 + data_f0_cast + data_f1_b6 + data_f1_cast

# collect genetic effects
results_per_gene_expressed_intervals <- readRDS("./Data/processed/gp_fits_DE_new.rds")
results_per_gene_expressed_intervals <- results_per_gene_expressed_intervals[!unlist(lapply(results_per_gene_expressed_intervals, function(x){any(is.na(x))}))]

likelihood_df <- data.frame(
  likelihood_cons = unlist(lapply(results_per_gene_expressed_intervals, function(x){x[[1]][[3]]})),
  likelihood_stat = unlist(lapply(results_per_gene_expressed_intervals, function(x){x[[2]][[3]]})),
  likelihood_dyn = unlist(lapply(results_per_gene_expressed_intervals, function(x){x[[3]][[3]]}))
)
bf_dynamic_deg <- likelihood_df$likelihood_dyn - likelihood_df$likelihood_stat
names(bf_dynamic_deg) <- names(results_per_gene_expressed_intervals)

genes_deg_use <- names(bf_dynamic_deg[bf_dynamic_deg > 5])
all_genes_show <- union(genes_cis_dynamic, union(genes_trans_dynamic, genes_deg_use))

# collect matrices of all effects

library(zoo)

total_expression_unscaled <- data_f0_b6 + data_f0_cast + data_f1_b6 + data_f1_cast
total_expression_unscaled <- total_expression_unscaled[rownames(genes_test), ]
# total_expression <- log(total_expression_unscaled + 1)
total_expression <- total_expression[rownames(genes_test), ]
total_expression <- data.frame(t(rollapply(t(total_expression), width = 10, by = 1, FUN = mean, align = "left")))
# total_expression_unscaled <- total_expression

rs_f0_smoothed <- data.frame(t(rollapply(t((rs_f0)), width = 10, by = 1, FUN = mean, align = "left")))
rs_f1_smoothed <- data.frame(t(rollapply(t((rs_f1)), width = 10, by = 1, FUN = mean, align = "left")))

rs_f0_smoothed_nooffset <- data.frame(t(rollapply(t((rs_f0_no_offset)), width = 10, by = 1, FUN = mean, align = "left")))
rs_f1_smoothed_nooffset <- data.frame(t(rollapply(t((rs_f1_no_offset)), width = 10, by = 1, FUN = mean, align = "left")))

total_expression <- total_expression
deg_effects <- rs_f0_smoothed
cis_effects <- rs_f1_smoothed
trans_effects <- rs_f0_smoothed - rs_f1_smoothed

scale_to_zero_one <- function(x){(x - min(x)) / (max(x) - min(x))}
# scale_to_zero_one <- function(x){(x - min(x))}

total_expression <- t(apply(total_expression, 1, scale_to_zero_one))

deg_effects <- deg_effects[intersect(genes_deg_use, rownames(deg_effects)),  ]
deg_effects <- abs(deg_effects - 0.5)
deg_effects_unscaled <- abs(rs_f0_smoothed_nooffset - 0.5)
deg_effects_unscaled <- deg_effects
deg_effects <- t(apply(deg_effects, 1, scale_to_zero_one))

cis_effects <- cis_effects[genes_cis_dynamic, ]
cis_effects <- abs(cis_effects - 0.5)
cis_effects_unscaled <- abs(rs_f1_smoothed_nooffset - 0.5)
cis_effects_unscaled <- cis_effects
cis_effects <- t(apply(cis_effects, 1, scale_to_zero_one))

trans_effects <- trans_effects[genes_trans_dynamic, ]
trans_effects <- abs(trans_effects)
trans_effects_unscaled <- abs(rs_f0_smoothed_nooffset - rs_f1_smoothed_nooffset)
trans_effects_unscaled <- trans_effects
trans_effects <- t(apply(trans_effects, 1, scale_to_zero_one))

n_clusters <- 5

all_genetic_effects <- data.frame(
  gene = all_genes_show, 
  cis = all_genes_show %in% genes_cis_dynamic, 
  trans = all_genes_show %in% genes_trans_dynamic, 
  deg = all_genes_show %in% genes_deg_use
)

# correlation BFs cis and deg

color_mapping = list(
  "no_effect" = "grey", 
  "only_cis" = "#DEC8B4", 
  "only_trans" = "#996737", 
  "both" = "#473422"
)

# get percentages
data.frame(table(all_genetic_effects[all_genetic_effects$deg, ][,c("cis", "trans")])) %>%
  add_column(new_column = paste0(.$cis, "_", .$trans)) %>%
  mutate(new_column = recode(new_column, !!!c("FALSE_FALSE" = "no_effect", "FALSE_TRUE" = "only_trans", "TRUE_FALSE" = "only_cis", "TRUE_TRUE" = "both"))) %>%
  mutate(new_column = factor(new_column, levels = c("no_effect", "only_cis", "only_trans", "both"))) %>%
  mutate(Freq = Freq / sum(Freq))

data.frame(table(all_genetic_effects[all_genetic_effects$deg, ][,c("cis", "trans")])) %>%
  add_column(new_column = paste0(.$cis, "_", .$trans)) %>%
  mutate(new_column = recode(new_column, !!!c("FALSE_FALSE" = "no_effect", "FALSE_TRUE" = "only_trans", "TRUE_FALSE" = "only_cis", "TRUE_TRUE" = "both"))) %>%
  mutate(new_column = factor(new_column, levels = c("no_effect", "only_cis", "only_trans", "both"))) %>%
  ggplot(aes(x = 1, y = Freq, fill = new_column)) + 
  geom_bar(stat = "identity", position = "stack", col = "black") + 
  theme_void() + scale_fill_manual(values = color_mapping) +
  theme(aspect.ratio = 3, text = element_text(size = 30))
ggsave("./Plots/Figure4/F4_clustering_barplot_cis_trans.pdf")

# total unscaled plots 
total_expression_unscaled_here <- total_expression_unscaled
deg_effects_unscaled_here <- as.matrix(deg_effects_unscaled[intersect(rownames(total_expression_unscaled), rownames(deg_effects_unscaled)), ])
cis_effects_unscaled_here <- as.matrix(cis_effects_unscaled[intersect(rownames(total_expression_unscaled), rownames(cis_effects_unscaled)), ])
trans_effects_unscaled_here <- as.matrix(trans_effects_unscaled[intersect(rownames(total_expression_unscaled), rownames(trans_effects_unscaled)), ])

total_expression_unscaled_here_na <- total_expression_unscaled_here < 0.5

deg_effects_unscaled_here <- do.call("rbind", lapply(rownames(deg_effects_unscaled_here), function(x){
  xx <- deg_effects_unscaled_here[x, ]
  xx[total_expression_unscaled_here_na[x, ]] <- NA
  xx
}))
cis_effects_unscaled_here <- do.call("rbind", lapply(rownames(cis_effects_unscaled_here), function(x){
  xx <- cis_effects_unscaled_here[x, ]
  xx[total_expression_unscaled_here_na[x, ]] <- NA
  xx
}))
trans_effects_unscaled_here <- do.call("rbind", lapply(rownames(trans_effects_unscaled_here), function(x){
  xx <- trans_effects_unscaled_here[x, ]
  xx[total_expression_unscaled_here_na[x, ]] <- NA
  xx
}))

total_effect_sizes_unscaled <- data.frame(
  "Bins" = 1:91, 
  "total" = colMeans(total_expression_unscaled)[1:91],
  "deg" = colMeans(deg_effects_unscaled),
  "cis" = colMeans(cis_effects_unscaled),
  "trans" = colMeans(trans_effects_unscaled)
) %>% pivot_longer(-c("Bins")) %>%
  mutate(name = factor(name, levels = c("total", "deg", "cis", "trans")))

ggplot(total_effect_sizes_unscaled[total_effect_sizes_unscaled$name == "total", ], 
       aes(x = Bins, y = value, col = name)) + 
  geom_smooth(span = 0.1, col = "black", linetype = 'dashed') + 
  geom_smooth(span = 1, col = "darkgrey", linetype = 'dashed', alpha = 0.3) + 
  # geom_point(alpha = 0.3, col = "black") + 
  theme_classic() + 
  theme(text = element_text(size = 30), 
        legend.position = "None") + 
  ylab("") + xlab("") + 
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_unscaled_expression.pdf", height = 5, width = 40, unit = "cm")

ggplot(total_effect_sizes_unscaled[total_effect_sizes_unscaled$name == "deg", ], 
       aes(x = Bins, y = value, col = name)) + 
  geom_smooth(span = 0.1, col = "black", linetype = 'dashed') + 
  geom_smooth(span = 1, col = "darkgrey", linetype = 'dashed', alpha = 0.3) + 
  # geom_point(alpha = 0.3, col = "black") + 
  theme_classic() + 
  theme(text = element_text(size = 30), 
        legend.position = "None") + 
  ylab("") + xlab("") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_differential_expression.pdf", height = 5, width = 40, unit = "cm")

ggplot(total_effect_sizes_unscaled[total_effect_sizes_unscaled$name == "cis", ], 
       aes(x = Bins, y = value, col = name)) + 
  geom_smooth(span = 0.1, col = "black", linetype = 'dashed') + 
  geom_smooth(span = 1, col = "darkgrey", linetype = 'dashed', alpha = 0.3) + 
  # geom_point(alpha = 0.3, col = "black") + 
  theme_classic() + 
  theme(text = element_text(size = 30), 
        legend.position = "None") + 
  ylab("") + xlab("") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_cis_effects.pdf", height = 5, width = 40, unit = "cm")

ggplot(total_effect_sizes_unscaled[total_effect_sizes_unscaled$name == "trans", ], 
       aes(x = Bins, y = value, col = name)) + 
  geom_smooth(span = 0.1, col = "black", linetype = 'dashed') + 
  geom_smooth(span = 1, col = "darkgrey", linetype = 'dashed', alpha = 0.3) + 
  # geom_point(alpha = 0.3, col = "black") + 
  theme_classic() + 
  theme(text = element_text(size = 30), 
        legend.position = "None") + 
  ylab("") + xlab("") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_trans_effects.pdf", height = 5, width = 40, unit = "cm")

###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### 
# joint clustering
###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### ###### 

make_meta_without_clustering <- function(dataset, clusters){
  cluster_counts <- table(clusters)
  res <- lapply(unique(clusters), function(i){
    dd <- dataset[clusters == i, ]
    if (  cluster_counts[[i]] == 1 ){
      return(dd)
    }
    return(colMeans(dd, na.rm = T))
  })
  return(do.call("rbind", res))
}

total_effects_here <- total_expression
deg_effects_here <- deg_effects
cis_effects_here <- cis_effects
trans_effects_here <- trans_effects

rownames(total_effects_here) <- paste0("TOTAL_", rownames(total_effects_here))
rownames(deg_effects_here) <- paste0("DEG_", rownames(deg_effects_here))
rownames(cis_effects_here) <- paste0("CIS_", rownames(cis_effects_here))
rownames(trans_effects_here) <- paste0("TRANS_", rownames(trans_effects_here))

joint_matrix <- rbind(
  total_effects_here, 
  deg_effects_here, 
  cis_effects_here, 
  trans_effects_here
)

set.seed(123)

total_clusters <- cutree(hclust(dist(joint_matrix), method = "complete"), k = 7)
table(total_clusters)
late_cluster <- c("3", "1", "5", "6")

view_annotation <- data.frame(
  Dummy = 1, 
  View = unlist(lapply(str_split(rownames(joint_matrix), "_"), function(x){x[[1]]})), 
  row.names = rownames(joint_matrix)
)

# look at cluster trajectories across different layers

trace_results <- lapply(1:7, function(i){
  dds = joint_matrix[total_clusters == i, ]
  total_trace = colMeans(dds[grepl("TOTAL", rownames(dds)), ])
  deg_trace = colMeans(dds[grepl("DEG", rownames(dds)), ])
  cis_trace = colMeans(dds[grepl("CIS", rownames(dds)), ])
  if (sum(grepl("TRANS", rownames(dds))) > 1){
    trans_trace = colMeans(dds[grepl("TRANS", rownames(dds)), ])
  }  
  if (sum(grepl("TRANS", rownames(dds))) == 1){
    trans_trace = dds[grepl("TRANS", rownames(dds)), ]
  }
  dds <- data.frame(do.call("rbind", list(total_trace, deg_trace, cis_trace, trans_trace)))
  dds$Cluster = paste0("Cluster_", i)
  dds$View = c("total", "deg", "cis", "trans")
  dds
})
trace_results <- data.frame(do.call("rbind", trace_results))

trace_results %>%
  pivot_longer(-c("Cluster", "View")) %>%
  mutate(name = as.numeric(gsub("X", "", name))) %>%
  ggplot(aes(x = name, y = value, col = View)) + geom_point() + 
  facet_wrap(~Cluster) + 
  theme_classic()

# how do seperate and joint clusterings align?
views_by_cluster <- unlist(lapply(str_split(names(total_clusters), "_"), function(x){x[[1]]}))
pheatmap(log10(table(views_by_cluster, total_clusters) + 1), cluster_rows = F, cluster_cols = F)

total_clusters_joint <- gsub("TOTAL_", "", names(total_clusters[grepl("TOTAL", names(total_clusters))]))
joint_sub_total <- total_clusters[grepl("TOTAL", names(total_clusters))]
deg_clusters_joint <- gsub("DEG_", "", names(total_clusters[grepl("DEG", names(total_clusters))]))
joint_sub_deg <- total_clusters[grepl("DEG", names(total_clusters))]
cis_clusters_joint <- gsub("CIS_", "", names(total_clusters[grepl("CIS", names(total_clusters))]))
joint_sub_cis <- total_clusters[grepl("CIS", names(total_clusters))]
trans_clusters_joint <- gsub("TRANS_", "", names(total_clusters[grepl("TRANS", names(total_clusters))]))
joint_sub_trans <- total_clusters[grepl("TRANS", names(total_clusters))]

cluster_df_comp <- data.frame(
  gene = rownames(total_expression), 
  #total_cluster_only = NA, 
  total_cluster_joint = NA, 
  #deg_cluster_only = NA, 
  deg_cluster_joint = NA, 
  #cis_cluster_only = NA, 
  cis_cluster_joint = NA, 
  #trans_cluster_only = NA, 
  trans_cluster_joint = NA
) %>% column_to_rownames("gene")

# cluster_df_comp[names(total_clusters_total), ]$total_cluster_only <- total_clusters_total
# cluster_df_comp[names(total_clusters_deg), ]$deg_cluster_only <- total_clusters_deg
# cluster_df_comp[names(total_clusters_cis), ]$cis_cluster_only <- total_clusters_cis
# cluster_df_comp[names(total_clusters_trans), ]$trans_cluster_only <- total_clusters_trans
cluster_df_comp[total_clusters_joint, ]$total_cluster_joint <- joint_sub_total
cluster_df_comp[deg_clusters_joint, ]$deg_cluster_joint <- joint_sub_deg
cluster_df_comp[cis_clusters_joint, ]$cis_cluster_joint <- joint_sub_cis
cluster_df_comp[trans_clusters_joint, ]$trans_cluster_joint <- joint_sub_trans

cluster_df_comp %>%
  dplyr::select(c(total_cluster_joint, deg_cluster_joint, cis_cluster_joint, trans_cluster_joint)) %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene) %>%
  ggplot(aes(value, fill = name)) + geom_bar(position = "fill") + 
  theme_classic()

cluster_df_comp %>%
  dplyr::select(c(total_cluster_joint, deg_cluster_joint, cis_cluster_joint, trans_cluster_joint)) %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene) %>%
  dplyr::filter(!is.na(value)) %>%
  ggplot(aes(name, fill = factor(value))) + geom_bar(position = "fill") + 
  theme_classic()

### colors and stuff
cluster_conversion <- c(
  "Late_1", # 1
  "Early_1", # 2
  "Late_4", # 3
  "Early_3", # 4
  "Late_2", # 5
  "Late_3", # 6
  "Early_2" # 7
)

color_1 <- "grey"
color_2 <- "#E69F00"
color_3 <- "#009E73"

cluster_colors <- c(
  "grey", 
  colorRampPalette(c(color_1, color_2, color_3))(7)[7], 
  "grey",
  colorRampPalette(c(color_1, color_2, color_3))(7)[6], 
  colorRampPalette(c(color_1, color_2, color_3))(7)[5], 
  "grey",
  colorRampPalette(c(color_1, color_2, color_3))(7)[4]
)
cluster_colors <- c(
  colorRampPalette(c(color_1, color_2, color_3))(7)[4],
  "grey",
  colorRampPalette(c(color_1, color_2, color_3))(7)[7], 
  "grey",
  colorRampPalette(c(color_1, color_2, color_3))(7)[5], 
  colorRampPalette(c(color_1, color_2, color_3))(7)[6], 
  "grey"
)
names(cluster_colors) <- as.character(1:7)

pheatmap(table(cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")]), cluster_rows = F, cluster_cols = F)
pheatmap(log10(table(cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")]) + 1), cluster_rows = F, cluster_cols = F, 
         col=c("white", "black"))

ttt <- cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")]
ttt <- ttt[apply(ttt, 1, function(x){all(!is.na(x))}), ]
# ttt <- ttt[apply(ttt, 1, function(x){all(x %in% c(1, 2, 3, 4, 5))}), ]

table(cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")]) %>%
  data.frame() %>%
  ggplot(aes(x = deg_cluster_joint, y = cis_cluster_joint, col = deg_cluster_joint, size = Freq)) + 
  geom_point() + 
  scale_color_manual(values = cluster_colors) + 
  theme_classic()  +
  xlab("DEG clusters") + 
  ylab("CIS clusters") + 
  theme(text = element_text(size = 30), 
        # legend.position = "None", 
        aspect.ratio = 1)

tt <- table(cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")])
tt <- tt / (outer(rowSums(tt), colSums(tt)) / sum(tt))

tt %>%
  data.frame() %>%
  ggplot() + 
  geom_point(aes(x = deg_cluster_joint, y = cis_cluster_joint, col = deg_cluster_joint, size = Freq)) + 
  # geom_point(data = data.frame(x = 1:7, y = 1:7), aes(x = x, y = y), col = "black", shape = 0, size = 10, stroke = 1) + 
  scale_color_manual(values = cluster_colors) + 
  theme_classic()  +
  xlab("DEG clusters") + 
  ylab("CIS clusters") + 
  theme(text = element_text(size = 20), 
        # legend.position = "None", 
        aspect.ratio = 1)
ggsave("./Plots/Figure4/F4_clusters_cis_deg.pdf", width = 5.3, height = 5.3)

tt <- cluster_df_comp[,c("deg_cluster_joint", "cis_cluster_joint")]
tt <- tt %>%
  data.frame() %>%
  dplyr::filter(!is.na(deg_cluster_joint) & !is.na(cis_cluster_joint))

library(ggsankey)

tt %>%
  data.frame() %>%
  dplyr::filter(!is.na(deg_cluster_joint) & !is.na(cis_cluster_joint)) %>%
  dplyr::filter(deg_cluster_joint %in% late_cluster & cis_cluster_joint %in% late_cluster) %>%
  make_long(deg_cluster_joint, cis_cluster_joint) %>%
  ggplot(aes(x = x, 
                 next_x = next_x, 
                 node = node, 
                 next_node = next_node,
                 fill = factor(node),
                 label = node)) +
    geom_sankey(flow.alpha = 0.5, node.color = 1) +
    geom_sankey_label(size = 3.5, color = 1, fill = "white") +
  scale_fill_manual(values = cluster_colors) + 
  theme_sankey(base_size = 16) + 
  theme(text = element_text(size = 20))
ggsave("./Plots/Figure4/F4_clusters_cis_deg_comparison.pdf", width = 6, height = 6)

# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 
# Total expression
# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 

testy <- trace_results %>% 
  dplyr::filter(View == "total")
plot_df <- testy %>% rownames_to_column() %>% pivot_longer(cols = -c(rowname, Cluster, View)) %>%
  mutate(name = as.numeric(gsub("X", "", name)))
dom_cluster <- c("3", "1", "5", "6")
plot_df$late_clusters <- ifelse(plot_df$rowname %in% dom_cluster, "late", "early")

ggplot() + 
  geom_smooth(data = plot_df, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("Pseudotime") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Total Expression") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/Supp_F4_clustering_TOTAL.pdf", width = 40, height = 10, units = "cm")

data.frame(
  Cluster = factor(1:length(unique(table(joint_sub_total))), levels = c("1","3", "5", "6", "4", "2", "7")), 
  nGenes = as.numeric(table(joint_sub_total))
) %>% ggplot(aes(x="", y = nGenes, fill = as.factor(Cluster))) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  coord_polar("y", start=0) + 
  scale_fill_manual(values = cluster_colors) + 
  theme_void()
ggsave("./Plots/Figure4/Supp_F4_clustering_TOTAL_venn.pdf")

tt <- table(joint_sub_total)
sum(tt[dom_cluster]) / sum(tt)

# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 
# DEG effects
# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 

testy <- trace_results %>% 
  dplyr::filter(View == "deg")
plot_df <- testy %>% rownames_to_column() %>% pivot_longer(cols = -c(rowname, Cluster, View)) %>%
  mutate(name = as.numeric(gsub("X", "", name)))
plot_df$late_clusters <- ifelse(plot_df$rowname %in% dom_cluster, "late", "early")

plot_df_here <- plot_df[plot_df$Cluster %in% c("Cluster_3", "Cluster_1", "Cluster_5", "Cluster_6"), ]

ggplot() + 
  geom_smooth(data = plot_df_here, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Differential Expression") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_clustering_DEG.pdf", width = 40, height = 12, units = "cm")

ggplot() + 
  geom_smooth(data = plot_df, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Differential Expression") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/Supp_F4_clustering_DEG.pdf", width = 40, height = 12, units = "cm")

data.frame(
  Cluster = factor(1:length(unique(table(joint_sub_deg))), levels = c("1","3", "5", "6", "4", "2", "7")), 
  nGenes = as.numeric(table(joint_sub_deg))
) %>% ggplot(aes(x="", y = nGenes, fill = as.factor(Cluster))) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  coord_polar("y", start=0) + 
  scale_fill_manual(values = cluster_colors) + 
  theme_void()
ggsave("./Plots/Figure4/F4_clustering_DEG_venn.pdf")
ggsave("./Plots/Figure4/Supp_F4_clustering_DEG_venn.pdf")

tt <- table(joint_sub_deg)
sum(tt[dom_cluster]) / sum(tt)

# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 
# Cis effects
# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 

testy <- trace_results %>% 
  dplyr::filter(View == "cis")
plot_df <- testy %>% rownames_to_column() %>% pivot_longer(cols = -c(rowname, Cluster, View)) %>%
  mutate(name = as.numeric(gsub("X", "", name)))
plot_df$late_clusters <- ifelse(plot_df$rowname %in% dom_cluster, "late", "early")

plot_df_here <- plot_df[plot_df$Cluster %in% c("Cluster_3", "Cluster_1", "Cluster_5", "Cluster_6"), ]

ggplot() + 
  geom_smooth(data = plot_df, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Cis Effects") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/Supp_F4_clustering_CIS.pdf", width = 40, height = 12, units = "cm")

ggplot() + 
  geom_smooth(data = plot_df_here, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Cis Effects") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_clustering_CIS.pdf", width = 40, height = 12, units = "cm")

data.frame(
  Cluster = factor(names(table(joint_sub_cis)), levels = c("1","3", "5", "6", "4", "2", "7")), 
  nGenes = as.numeric(table(joint_sub_cis))
) %>% ggplot(aes(x="", y = nGenes, fill = as.factor(Cluster))) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  coord_polar("y", start=0) + 
  scale_fill_manual(values = cluster_colors) + 
  theme_void()
ggsave("./Plots/Figure4/F4_clustering_CIS_venn.pdf")
ggsave("./Plots/Figure4/Supp_F4_clustering_CIS_venn.pdf")

tt <- table(joint_sub_cis)
sum(tt[dom_cluster]) / sum(tt)

# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 
# Trans effects
# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 

testy <- trace_results %>% 
  dplyr::filter(View == "trans")
plot_df <- testy %>% rownames_to_column() %>% pivot_longer(cols = -c(rowname, Cluster, View)) %>%
  mutate(name = as.numeric(gsub("X", "", name)))
plot_df$late_clusters <- ifelse(plot_df$rowname %in% dom_cluster, "late", "early")

plot_df_here <- plot_df[plot_df$Cluster %in% c("Cluster_3", "Cluster_1", "Cluster_5", "Cluster_6"), ]

ggplot() + 
  geom_smooth(data = plot_df, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Trans Effects") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/Supp_F4_clustering_TRANS.pdf", width = 40, height = 12, units = "cm")

ggplot() + 
  geom_smooth(data = plot_df_here, span = 0.5, size = 2, aes(group = rowname, x = name, y = value, col = rowname)) + 
  theme_classic() +
  theme(text = element_text(size = 30)) + 
  xlab("") + 
  ylab("Relative effect size") + 
  theme(legend.position = "None") + 
  scale_color_manual(values = cluster_colors) + 
  ggtitle("Trans Effects") + 
  theme(axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1))
ggsave("./Plots/Figure4/F4_clustering_TRANS.pdf", width = 40, height = 12, units = "cm")

data.frame(
  Cluster = factor(names(table(joint_sub_trans)), levels =  c("1","3", "5", "6", "4", "2", "7")), 
  nGenes = as.numeric(table(joint_sub_trans))
) %>% ggplot(aes(x="", y = nGenes, fill = as.factor(Cluster))) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  coord_polar("y", start=0) + 
  scale_fill_manual(values = cluster_colors) + 
  theme_void()
ggsave("./Plots/Figure4/F4_clustering_TRANS_venn.pdf")
ggsave("./Plots/Figure4/Supp_F4_clustering_TRANS_venn.pdf")

tt <- table(joint_sub_trans)
sum(tt[dom_cluster]) / sum(tt)

# ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## ######## 

colour_vector <- c("SG" = "#046735",
                   "SC" = "#D31281",
                   "RS" = "#282B69",
                   "ES" = "grey60",
                   "Sertoli" = "#643F18",
                   "Leydig" = "#985D25",
                   "Immune" = "#FFDFB2")

pt_here <- data_use$Pseudotime

data.frame(
  Pseudotime = pt_here, 
  CellType = data_use$CellType
) %>%
  ggplot(aes(x = Pseudotime, y = 1, col = CellType)) + geom_jitter(alpha = 1, height = 0.1) + 
  theme_classic() + ylim(0, 2) + 
  ylab("") + 
  xlab("") + 
  theme(legend.position = "None", 
        text = element_text(size = 30), 
        axis.line.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.text.y = element_blank()) + 
  scale_color_manual(values = colour_vector)
ggsave("./Plots/Figure4/Pseudotime_Plot.pdf")

###

```
