---
title: "Preprocessing of cross-species dataset"
author: "Nils Eling / Jasper Panten"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script reads in the cellranger v3 output of each 10X library, joins the objects and performs quality control as well as further downstream analysis tasks (normalization, dimensionality reduction, clustering).

# Read in data

```{r load_libraries, message=FALSE}
# Load libraries
library(DropletUtils)
library(scran)
library(reshape2)
library(scater)
library(ggplot2)
library(Rtsne)
library(irlba)
library(RColorBrewer)
library(viridis)
library(pheatmap)
library(batchelor)
library(umap)
library(tidyverse)
source("~/Desktop/PhD/RCode/Multipurpose/auxiliary.R")

setwd("~/Desktop/Projects/ASE_Spermatogenesis_Paper/")
```

```{r read_data, message=FALSE}
# Read in data - one SingleCellExperiment object per batch
# B6 
# Young
sce.do17815 <- read10xCounts("./Data/raw/evo_dataset/Sample1_B6/")
colData(sce.do17815)$Sample <- rep("B6", ncol(sce.do17815))
colData(sce.do17815)$Library <- rep("do17815", ncol(sce.do17815))
colData(sce.do17815)$Age <- rep("Young", ncol(sce.do17815))

sce.do17816 <- read10xCounts("./Data/raw/evo_dataset/Sample2_B6/")
colData(sce.do17816)$Sample <- rep("B6", ncol(sce.do17816))
colData(sce.do17816)$Library <- rep("do17816", ncol(sce.do17816))
colData(sce.do17816)$Age <- rep("Young", ncol(sce.do17816))

# CAST 
# Young
sce.do17811 <- read10xCounts("./Data/raw/evo_dataset/Sample1_CAST/")
colData(sce.do17811)$Sample <- rep("CAST", ncol(sce.do17811))
colData(sce.do17811)$Library <- rep("do17811", ncol(sce.do17811))
colData(sce.do17811)$Age <- rep("Young", ncol(sce.do17811))

sce.do17812 <- read10xCounts("./Data/raw/evo_dataset/Sample2_CAST/")
colData(sce.do17812)$Sample <- rep("CAST", ncol(sce.do17812))
colData(sce.do17812)$Library <- rep("do17812", ncol(sce.do17812))
colData(sce.do17812)$Age <- rep("Young", ncol(sce.do17812))

# Caroli 
# Young
sce.do17819 <- read10xCounts("./Data/raw/evo_dataset/Sample1_CAROLI/")
colData(sce.do17819)$Sample <- rep("CAROLI", ncol(sce.do17819))
colData(sce.do17819)$Library <- rep("do17819", ncol(sce.do17819))
colData(sce.do17819)$Age <- rep("Young", ncol(sce.do17819))

sce.do17820 <-  read10xCounts("./Data/raw/evo_dataset/Sample2_CAROLI/")
colData(sce.do17820)$Sample <- rep("CAROLI", ncol(sce.do17820))
colData(sce.do17820)$Library <- rep("do17820", ncol(sce.do17820))
colData(sce.do17820)$Age <- rep("Young", ncol(sce.do17820))
```

Next, we will merge the datasets based on their gene annotation.
We assume genes with Symbols that do not contain "MGP_" can be compared across species.

```{r common-genes}
# Find intersection
common_genes <- intersect(intersect(rowData(sce.do17815)$Symbol,
                                    rowData(sce.do17811)$Symbol),
                                    rowData(sce.do17819)$Symbol)
```

We retrieve 16,732 genes that are shared across the species.
We first merge the species-specific datasets and subset based on their genes.

```{r}
# B6
B6.all <- cbind(sce.do17815, sce.do17816)
B6.all <- B6.all[match(common_genes, rowData(B6.all)$Symbol),]
rownames(B6.all) <- rowData(B6.all)$Symbol
rowData(B6.all) <- rowData(B6.all)[,2]

# CAST
CAST.all <- cbind(sce.do17811, sce.do17812)
CAST.all <- CAST.all[match(common_genes, rowData(CAST.all)$Symbol),]
rownames(CAST.all) <- rowData(CAST.all)$Symbol
rowData(CAST.all) <- rowData(CAST.all)[,2]

# CAROLI
CAROLI.all <- cbind(sce.do17819, sce.do17820)
CAROLI.all <- CAROLI.all[match(common_genes, rowData(CAROLI.all)$Symbol),]
rownames(CAROLI.all) <- rowData(CAROLI.all)$Symbol
rowData(CAROLI.all) <- rowData(CAROLI.all)[,2]

# Merge all cells
sce.all <- cbind(B6.all, CAST.all, CAROLI.all)
colnames(sce.all) <- paste0(sce.all$Library, "_", sce.all$Barcode)

rm(list = c("B6.all", "CAST.all", "CAROLI.all"))
rm(list = c("sce.do17815", "sce.do17816", "sce.do17811", "sce.do17812", "sce.do17819", "sce.do17820"))
```

# Filtering 

```{r Filtering}
# Save statistics on these libraries
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- data.frame(row.names = cur_stats$Var2,
                       Sample = cur_stats$Var1,
                       Library = cur_stats$Var2,
                       No_cells = cur_stats$value)

# Calculate QC metrics
colData(sce.all) <- cbind(colData(sce.all), perCellQCMetrics(sce.all))

sce.all <- sce.all[,colData(sce.all)$sum > 500]
sce.all <- sce.all[,colData(sce.all)$detected > 500]

# Remove genes that are not expressed
sce.all <- sce.all[Matrix::rowSums(counts(sce.all)) > 0,]

# Add to stats data frame
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df$AfterFiltering <- cur_stats$value

write.csv(stats.df, "./Data/processed/filtering_stats_evo.csv")
```

# Normalization

We next normalize the data based on the pooling method inplemented in scran.

```{r normalization}
clusters <- quickCluster(sce.all, method = "igraph",
                         use.ranks=FALSE, min.size = 100)

sce.all <- computeSumFactors(sce.all, clusters=clusters)
sce.all <- logNormCounts(sce.all)
```

# Quality visualization

```{r plotting}
# Highly variable genes

HVgenes <- HVG(sce = sce.all)

set.seed(12345)
pca <- prcomp_irlba(t(logcounts(sce.all[HVgenes,])), n = 50)
tsne <- Rtsne(pca$x, pca = FALSE)
reducedDims(sce.all)$TSNE <- tsne$Y

# Batches
batches <- unique(paste(colData(sce.all)$Sample, colData(sce.all)$Library))
col_vector_batch <- vector(length = length(batches))
names(col_vector_batch) <- batches
col_vector_batch["B6 do17815"] <- colorRampPalette(c("white", "black"))(10)[10]
col_vector_batch["B6 do17816"] <- colorRampPalette(c("white", "black"))(10)[8]
col_vector_batch["CAST do17811"] <- colorRampPalette(c("white", "chocolate"))(10)[10]
col_vector_batch["CAST do17812"] <- colorRampPalette(c("white", "chocolate"))(10)[8]
col_vector_batch["CAROLI do17819"] <- colorRampPalette(c("white", "brown4"))(10)[4]
col_vector_batch["CAROLI do17820"] <- colorRampPalette(c("white", "brown4"))(10)[2]

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  batch = paste(colData(sce.all)$Sample, 
                                colData(sce.all)$Library))) +
  		  geom_point(aes(tsne1, tsne2, colour = batch)) +
  		  scale_color_manual(values = col_vector_batch)

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$detected)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$detected)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Some marker genes
# Stra8 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Stra8",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Defb19 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Defb19",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Prm1 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Prm1",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Dazl 
  ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Dazl",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

```

# Batch correction

```{r correction}
# Split datasets into batches
sce.single <- split.sce(sce = sce.all, groups = unique(colData(sce.all)$Library), 
                        colData.name = "Library")

# Batch correction
corrected <- batch.correction(sce.single)
#rm(sce.single)

# Save batch corrected matrix in sce object
metadata(sce.all)$corrected <- corrected

corrected <- as.matrix(corrected)

tsne <- Rtsne(corrected, pca = FALSE)
set.seed(12345)
reducedDims(sce.all)$TSNE <- tsne$Y
set.seed(12345)
umap <- umap::umap(corrected, pca = FALSE)
reducedDims(sce.all)$UMAP <- umap$layout

```

# Visualize batch corrected data

```{r visualizataion}
# Batches
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  batch = paste(colData(sce.all)$Sample,
                                colData(sce.all)$Library))) +
  geom_point(aes(tsne1, tsne2, colour = batch)) +
  scale_color_manual(values = col_vector_batch)

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$detected)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Plot gene expression
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Stra8",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Hormad1",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Cst12",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Insl3",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Tm4sf1",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Pou5f2",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)["Prm1",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

```

# Clustering

We will now cluster the batch corrected data.

```{r}
set.seed(1234)
g <- buildSNNGraph(t(metadata(sce.all)$corrected), k = 4)
clusters <- igraph::cluster_louvain(g)$membership
table(clusters)

# Visualize clustering
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_cluster = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  cluster = as.factor(clusters))) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = col_vector_cluster)

ggplot(data.frame(tsne1 = reducedDims(sce.all)$UMAP[,1],
                  tsne2 = reducedDims(sce.all)$UMAP[,2],
                  cluster = as.factor(clusters))) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = col_vector_cluster)

colData(sce.all)$Clusters <- clusters
```

# Compare current with previous cluster annotation

In the previous publication (Ernst et al, 2019, Nature Communications), we performed in-depth profiling of spermatogenesis cell-types.
Since two of the libraries are used again for this project, we will match the cell-types of the previous publication with the current data.
The metadata can be obtained from ArrayExpress.

```{r cell-type-recognition}
# Download metadata file
#download.file("https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-6946/E-MTAB-6946.processed.3.zip", 
#                destfile = "cellranger_metadata_1.zip")
#download.file("https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-6946/E-MTAB-6946.processed.1.zip", 
#                destfile = "cellranger_metadata.zip")
#download.file("https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-6946/E-MTAB-6946.processed.5.zip", 
#                destfile = "cellranger_metadata_emptyDrops.zip")
#unzip("cellranger_metadata.zip") #
#unzip("cellranger_metadata.zip") #
#unzip("cellranger_metadata_emptyDrops.zip") #

#file.remove("cellranger_metadata.zip")
#file.remove("cellranger_metadata_emptyDrops.zip")

#metadata <- read.table("./cell_metadata.txt", sep = " ")
metadata <- read.table("./cell_metadata_emptyDrops.txt", sep = " ")

# Select B6 cells
metadata <- metadata[metadata$Sample == "B6",]
cur_colData <- colData(sce.all)[sce.all$Sample == "B6",]

# Match cell identifiers between both dataset
cur_ID <- intersect(paste(metadata$Barcode, metadata$Library, sep = "_"),
                    paste(cur_colData$Barcode, cur_colData$Library, sep = "_"))
metadata <- metadata[match(cur_ID,
                           paste(metadata$Barcode, metadata$Library, sep = "_")),]
cur_colData <- cur_colData[match(cur_ID,
                           paste(cur_colData$Barcode, cur_colData$Library, sep = "_")),]

# Visualize the overlap of clusters and cell-types
mat <- table(cur_colData$Clusters, metadata$AnnotatedClusters)
mat_notnew <- mat[,colnames(mat) != "NewCell"]
mapping_to_old <- apply(mat_notnew, 1, function(x){colnames(mat_notnew)[which(x == max(x))][[1]]})

pheatmap(log(mat + 1), cluster_cols = F, cluster_rows = F)

```

# Cell-type annotation

We detect some mixed cell-types and will now label the clusters by hand.

```{r cluster-labelling}

# Look at liefted annotation in dataset
clusters = sce.all$Clusters

data.frame(
  umap1 = reducedDims(sce.all)[["UMAP"]][,1], 
  umap2 = reducedDims(sce.all)[["UMAP"]][,2],
  celltype_lifted = mapping_to_old[clusters]
) %>%
  ggplot(aes(umap1, umap2, col = celltype_lifted)) + 
  	geom_point()

# Jasper new annotation with new clustering
new.vector <- clusters
new.vector[new.vector == 1] <- "Elongating Spermatids (4)"
new.vector[new.vector == 2] <- "Elongating Spermatids (6)"
new.vector[new.vector == 3] <- "Outliers" #
new.vector[new.vector == 4] <- "Sertoli"
new.vector[new.vector == 5] <- "Outliers"
new.vector[new.vector == 6] <- "Round Spermatids (3)" #
new.vector[new.vector == 7] <- "Spermatogonia" # 
new.vector[new.vector == 8] <- "Immune"
new.vector[new.vector == 9] <- "Elongating Spermatids (2)"
new.vector[new.vector == 10] <- "Elongating Spermatids (1)"
new.vector[new.vector == 11] <- "Elongating Spermatids (10)"
new.vector[new.vector == 12] <- "Outliers"
new.vector[new.vector == 13] <- "Pachytene" #
new.vector[new.vector == 14] <- "Elongating Spermatids (7)"
new.vector[new.vector == 15] <- "Elongating Spermatids (5)"
new.vector[new.vector == 16] <- "Elongating Spermatids (9)"
new.vector[new.vector == 17] <- "Pachytene"
new.vector[new.vector == 18] <- "Elongating Spermatids (8)"
new.vector[new.vector == 19] <- "Elongating Spermatids (8)"
new.vector[new.vector == 20] <- "Elongating Spermatids (8)"
new.vector[new.vector == 21] <- "Elongating Spermatids (10)"
new.vector[new.vector == 22] <- "Round Spermatids (4)" #
new.vector[new.vector == 23] <- "Meiosis"
new.vector[new.vector == 24] <- "Round Spermatids (1)" #
new.vector[new.vector == 25] <- "Diplotene"
new.vector[new.vector == 26] <- "Leydig" 
new.vector[new.vector == 27] <- "Round Spermatids (5)"
new.vector[new.vector == 28] <- "Round Spermatids (2)" #
new.vector[new.vector == 29] <- "Outliers"
new.vector[new.vector == 30] <- "Elongating Spermatids (3)"

colour_vector <- c("Spermatogonia" = "#046735",
                   "Pachytene" = "#D31281",
                   "Diplotene" = "#7E2678",
                   "Meiosis" = "#502269",
                   "Round Spermatids (1)" = "#B3A9D3",
                   "Round Spermatids (2)" = "#D8DAEC",
                   "Round Spermatids (3)" = "#9595C9",
                   "Round Spermatids (4)" = "#6167AF",
                   "Round Spermatids (5)" = "#2A348B",
                   "Elongating Spermatids (1)" = "#171447",
                   "Elongating Spermatids (2)" = "#22232B",
                   "Elongating Spermatids (3)" = "#444A5E",
                   "Elongating Spermatids (4)" = "#606B89",
                   "Elongating Spermatids (5)" = "#737884",
                   "Elongating Spermatids (6)" = "grey50",
                   "Elongating Spermatids (7)" = "grey60",
                   "Elongating Spermatids (8)" = "grey70",
                   "Elongating Spermatids (9)" = "grey80",
                   "Elongating Spermatids (10)" = "grey90",
                   "Sertoli" = "#643F18",
                   "Leydig" = "#985D25",
                   "Immune" = "#C48C58")

sce.all.clean <- sce.all[, new.vector != "Outliers"]
metadata(sce.all.clean)$corrected <- metadata(sce.all.clean)$corrected[colnames(sce.all.clean), ]

sce.all.clean$CellType <- factor(new.vector[new.vector != "Outliers"], levels = names(colour_vector))

```

Recompute umap after outlier exclusion

```{r recompute_tsne}

set.seed(12345)
umap <- umap::umap(metadata(sce.all.clean)$corrected, pca = FALSE)
reducedDims(sce.all.clean)$UMAP <- umap$layout

ggplot(data.frame(Tsne1 = reducedDims(sce.all.clean)$UMAP [,1],
                  Tsne2 = reducedDims(sce.all.clean)$UMAP [,2],
                  cluster = sce.all.clean$CellType)) +
        geom_point(aes(Tsne1, Tsne2, colour = cluster)) +
        scale_color_manual(values = colour_vector)

saveRDS(sce.all.clean, "./data/processed/final_sce_evo_dataset.rds")

```

